### PR TITLE
Bluetooth: controller: Handle AD data set race condition

### LINKED ISSUE #27637: Bluetooth: controller: Possible corruption in AD data
**Describe the bug**
On review of the code blocks below, the intended design requirement was to be able to supply AD data between HCI layer to Lower Link Layer (LLL) ISR in a context safe way, which is not the case.

Implementation used in the HCI context, to fill a buffer:
```
static inline struct pdu_adv *lll_adv_pdu_alloc(struct lll_adv_pdu *pdu,
						uint8_t *idx)
{
	uint8_t last;

	if (pdu->first == pdu->last) {
		last = pdu->last + 1;
		if (last == DOUBLE_BUFFER_SIZE) {
			last = 0U;
		}
	} else {
		last = pdu->last;
	}

	*idx = last;

	return (void *)pdu->pdu[last];
}
```

Implementation used in ISR to read a buffer:
```
static inline struct pdu_adv *lll_adv_pdu_latest_get(struct lll_adv_pdu *pdu,
						     uint8_t *is_modified)
{
	uint8_t first;

	first = pdu->first;
	if (first != pdu->last) {
		first += 1U;
		if (first == DOUBLE_BUFFER_SIZE) {
			first = 0U;
		}
		pdu->first = first;
		*is_modified = 1U;
	}

	return (void *)pdu->pdu[first];
}
```

If the HCI layer updates two or more times before ISR is run, then under race conditions, the ISR running while the thread mode context is modifying the `last` indexed buffer, (ISR) could be reading corrupt buffer content.

**To Reproduce**
Based on code review, it is believed the issue can be reproduced by changing the AD data as fast as possible iteratively while the advertising is already enabled.

**Expected behavior**
Contents of the AD Data transmitted over-air remains correct at all times.

**Impact**
showstopper for use cases where AD data is changed while advertising is enabled.

**Logs and console output**
NA

**Environment (please complete the following information):**
 - OS: (e.g. Linux, MacOS, Windows)
 - Toolchain (e.g Zephyr SDK, ...)
 - Commit SHA or Version used

**Additional context**

### PR DESCRIPTION
**Pull request by cvinayak**
Detect and handle AD data set race condition between thread
and ISR context.

Fixes #27637.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

### CODE REVIEW

**Comment by cvinayak** [comment]
@kruithofa To reproduce the race condition, build the added tests/bluetooth/adv application for nRF52840dk without the fix commit and flash it. Few seconds of running on the board will throw assertion in the UART console. (As the advertiser uses random delay, the reproduction could take some time).

The commits are in the order of reproducing the race condition and final commit is to handle the race condition.

My intention is to revert the 3 commits used to reproduce the race condition once reviewers are able to reproduce the issue and verify the fix.

@jhedberg What is your advice regarding test code used to reproduce an issue on the target? As the added test code will not be exercised in CI nor required by users, the test code (commits) be reverted, the commits be merge with history maintained?

File: tests/bluetooth/adv/src/main.c [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+	err = bt_enable(NULL);
+	if (err) {
+		printk("Bluetooth init failed (err %d)\n", err);
+	}
+	printk("Bluetooth initialized\n");
+
+	/* Start advertising */
+	err = bt_le_adv_start(BT_LE_ADV_NCONN, ad, ARRAY_SIZE(ad),
+			      sd, ARRAY_SIZE(sd));
+	if (err) {
```
Comment (kruithofa) [inline tests/bluetooth/adv/src/main.c:43]: replace with 'zassert_equal(err, 0, "Advertising failed to start (err %d)", err);

File: tests/bluetooth/adv/src/main.c [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+void test_adv_fast_ad_data_update(void)
+{
+	int err;
+
+	printk("Starting Beacon Demo\n");
+
+	/* Initialize the Bluetooth Subsystem */
+	err = bt_enable(NULL);
+	if (err) {
```
Comment (kruithofa) [inline tests/bluetooth/adv/src/main.c:35]: replace with 'zassert_equal(err, 0, "Bluetooth init failed (err %d)",err);

**Review by kruithofa — CHANGES_REQUESTED** [review]

**Review by kruithofa — APPROVED** [review]
LGTM

**Review by carlescufi — APPROVED** [review]

**Pull request merged**